pkg/util/resource.go

```diff
@@ -168,17 +168,26 @@ func (r *Resource) SetScalar(name corev1.ResourceName, quantity int64) {
 
 // ResourceList returns a resource list of this resource.
 func (r *Resource) ResourceList() corev1.ResourceList {
-	result := corev1.ResourceList{
-		corev1.ResourceCPU:              *resource.NewMilliQuantity(r.MilliCPU, resource.DecimalSI),
-		corev1.ResourceMemory:           *resource.NewQuantity(r.Memory, resource.BinarySI),
-		corev1.ResourceEphemeralStorage: *resource.NewQuantity(r.EphemeralStorage, resource.BinarySI),
-		corev1.ResourcePods:             *resource.NewQuantity(r.AllowedPodNumber, resource.DecimalSI),
+	result := corev1.ResourceList{}
```

Comment:
Here initialized an empty map which length is `0` and not `nil`.
This result to `replicaRequirements: {}`:
```yaml
  replicaRequirements: {}
  replicas: 2
```
I suppose if no items in the map, just return nil.

Reply:
Hi @RainbowMango, I have some concerns.

1. If we return a nil map, some field, like `cluster.spec.resourceSummary.allocated` may not be shown in some cases when there is no pending pods in clusters. It's a little weird. Meanwhile, it will have some effects on general estimator.
2. It may cause a nil pointer error when calling `estimator.MaxAvailableReplicas`.

Reply:
> If we return a nil map, some field, like cluster.spec.resourceSummary.allocated may not be shown in some cases when there is no pending pods in clusters. It's a little weird.

I guess you mean `cluster.status.allocating` as you are saying pending pods.
In that case, current behavior is : 
```yaml
  resourceSummary:
    allocatable:
      cpu: "4"
      ephemeral-storage: 123722284Ki
      hugepages-1Gi: "0"
      hugepages-2Mi: "0"
      memory: 16265864Ki
      pods: "110"
    allocated:
      cpu: 950m
      ephemeral-storage: 100Mi
      memory: 290Mi
      pods: "9"
    allocating:
      cpu: "0"
      ephemeral-storage: "0"
      memory: "0"
      pods: "0"
```

If return nil here, it might look like:
```yaml
  resourceSummary:
    allocatable:
      cpu: "4"
      ephemeral-storage: 123722284Ki
      memory: 16265864Ki
      pods: "110"
    allocated:
      cpu: 950m
      ephemeral-storage: 100Mi
      memory: 290Mi
      pods: "9"
```
Do you mean this looks weird?

Reply:
That's right.

Reply:
> Meanwhile, it will have some effects on general estimator.

What is it?

Reply:
May have nil pointer error of `replicaRequirements`.
https://github.com/karmada-io/karmada/blob/master/pkg/estimator/client/general.go#L49-L50
https://github.com/karmada-io/karmada/blob/master/pkg/estimator/client/accurate.go#L41-L41

Reply:
For the general estimator:
```golang
	for key, value := range replicaRequirements.ResourceRequest {
		requestedQuantity := value.Value()
```
If the `replicaRequirements.ResourceRequest` is nil, then non-ops. if the `replicaRequirements` is nil, then panic.
So, what kind of error we are talking about?

For the accurate estimator: 
```golang
return se.maxAvailableReplicas(ctx, cluster, replicaRequirements.DeepCopy())
```
Same as a general estimator.

Seems both of them could be solved technically.

Reply:
I didn't mean to change the behavior of the estimator and the Cluster. But I'm glad that you point it out. Then we can discuss it. No objection, just clarify your concerns.

Reply:
I mean the latter one. It is not a big problem and it is easy to fix it. I'm willing to make a change.